File: LeetCode/50.kt
fun myPow(x: Double, n: Int): Double {
    if(n == 0)
        return 1.0
    var x = x
    var halfN = n / 2
    if(n < 0) {
        x = 1 / x
        halfN *= -1
    }
    var exp: Double = myPow(x, halfN)
    if(n % 2 == 0) return exp * exp
    else return exp * exp * x
}

fun main() {
    println(myPow(2.00000, -2147483648))
}

File: LeetCode/50.cc
class Solution {
public:
    double myPow(double x, int n) {
        if(n == 0) return 1;
        auto sq_pow = n < 0 ? myPow(1/x, -(n >> 1)) : myPow(x, n >> 1);
        auto pow = sq_pow * sq_pow;
        return n & 1 ? pow * x : pow;
    }
};